File: test/clock/clock.windows.ml
external init_time : unit -> unit = "clock_windows_init"

external get_time : unit -> (int64[@unboxed])
  = "undefined" "clock_windows_get_time"
  [@@noalloc]

let () = init_time ()

let now () =
  let value = get_time () in
  Int64.to_float value /. 1000000000.

File: test/clock/clock.windows.c
#include <windows.h>

#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/alloc.h>
#include <caml/fail.h>

#ifndef __unused
#define __unused(x) x __attribute((unused))
#endif

#define __unit() value __unused(unit)

static LARGE_INTEGER frequency;

CAMLprim value
clock_windows_init(__unit ()) {
  QueryPerformanceFrequence(&frequency);
  frequency.QuadPart = 1000000000L / frequency.QuadPart;

  return Val_unit;
}

uint64_t
clock_windows_get_time(__unit ()) {
  LARGE_INTEGER now;

  QueryPerformanceCounter(&now);

  return (now.QuadPart * frequency.QuadPart);
}
